feat: add file and url validation for FSIM parameters

djach7 opened the pull request:
Adds best-effort validation for FSIM parameters during parsing, enabling FSIMs to fail earlier with more targeted error messages. Specifically verifying the following:
- The wget parameter is a syntactically valid URL using http or https
- The download parameter is a file(s) that exists and is readable
- The upload parameter is a directory that exists and is writeable

djach7 commented:
/gemini review

runcom commented on cmd/owner.go:
```diff
@@ -213,6 +215,90 @@ func getOwnerServerState(config *OwnerServerConfig) (*OwnerServerState, error) {
+		if err != nil {
+			if os.IsNotExist(err) {
+				return fmt.Errorf("upload directory %q does not exist", uploadDir)
+			}
```
you can drop this check, the error gives you already the context like "file not found"

djach7 replied:
Removed in update

djach7 commented on cmd/owner.go:
```diff
@@ -213,6 +215,90 @@ func getOwnerServerState(config *OwnerServerConfig) (*OwnerServerState, error) {
+			}
+			if os.IsPermission(err) {
+				return fmt.Errorf("download file %q is not readable", filePath)
+			}
```
Removed in update

djach7 commented on cmd/owner.go:
```diff
@@ -213,6 +215,90 @@ func getOwnerServerState(config *OwnerServerConfig) (*OwnerServerState, error) {
+	downloadPaths = make([]string, 0, len(downloads))
+	for _, filePath := range downloads {
+		cleanPath := filepath.Clean(filePath)
+		file, err := os.Open(cleanPath)
```
I changed it to a stat barring any further discussion

kgiusti replied:
Turns out you can os.Stat() a file that you don't have read permission for, so we can't rely on a successful call to os.Stat() to assume readability.  Still have to open it I guess.

djach7 commented on cmd/owner.go:
```diff
@@ -213,6 +215,90 @@ func getOwnerServerState(config *OwnerServerConfig) (*OwnerServerState, error) {
+		if err != nil {
+			return fmt.Errorf("upload directory %q is not writable: %w", uploadDir, err)
+		}
+		testFileName := testFile.Name()
```
Changed in update

kgiusti commented:
@djach7 It would be useful to run the validation when the server comes up, since that's when configuration errors are more likely to be seen and can be corrected. The PR currently runs the validation at the start of onboarding a device which at that point is too late.  There's an OwnerServerConfig.validate() function which would be a good place to call the validation function.

Of course the files/directories may become invalid in the time between starting the server and onboarding. Meh - you mess with stuff at run time you get to keep the pieces...

djach7 commented:
That was my original implementation @kgiusti, I changed it to adapt to how the CI tests are setup. When the validation occurred at server startup then the CI (fsim download specifically, if I remember right) would fail, as the files don't exist yet when the servers are started up for the CI. I was assuming that the CI setup was going to be the predominate use case, and thus moved the validation to be at onboarding instead. 

Should I change it back and adapt the CI tests to fit the validation instead?

kgiusti commented:
> That was my original implementation @kgiusti, I changed it to adapt to how the CI tests are setup. When the validation occurred at server startup then the CI (fsim download specifically, if I remember right) would fail, as the files don't exist yet when the servers are started up for the CI. I was assuming that the CI setup was going to be the predominate use case, and thus moved the validation to be at onboarding instead.
> 
> Should I change it back and adapt the CI tests to fit the validation instead?

Yeah the tests should put everything in place before starting the server. That MNSHO.

djach7 commented:
Updated with the following:
- Moved the validation back to server startup
- Updated the fsim download CI test to create files before server startup
- `make` removed some trailing whitespace in `internal/db/db_owner_rvinfo_errors_test.go`

runcom commented:
Needs a rebase

runcom commented on cmd/owner.go:
```diff
@@ -213,6 +221,70 @@ func getOwnerServerState(config *OwnerServerConfig) (*OwnerServerState, error) {
+		testFile, err := os.CreateTemp(uploadDir, ".fdo-write-test-*")
+		if err != nil {
+			return fmt.Errorf("upload directory %q is not writable: %w", uploadDir, err)
+		}
```
I guess since this is just best effort checking - after the error check you can just do:
```
testFile.Close()
os.Remove(testFile.Name())
```
